lib/Backends/OpenCL/OpenCL.cpp

```diff
@@ -166,6 +178,100 @@ cl_kernel OpenCLFunction::createKernel(const std::string &name,
+  std::string programPath = cacheDirectory + '/' + programFileName;
+  if (!llvm::sys::fs::exists(programPath))
+    return nullptr;
+
+  uint64_t binarySize;
+  std::error_code errc = llvm::sys::fs::file_size(programPath, binarySize);
+  CHECK(!errc) << "Error getting the file size of " << programPath << ".";
+
+  int fileDesc = open(programPath.c_str(), O_RDONLY);
+  CHECK_GT(fileDesc, -1) << "Error opening " << programPath << " for reading.";
+
+  unsigned char *binary = new unsigned char[binarySize];
+  ssize_t readBytes = read(fileDesc, binary, binarySize);
+  close(fileDesc);
+  CHECK_EQ(readBytes, binarySize) << "Could not read the binary.";
+
+  cl_int status;
+  cl_int err;
+  cl_program prog =
+      clCreateProgramWithBinary(ctx, 1, &device, &binarySize,
+                                (const unsigned char **)&binary, &status, &err);
+  delete[] binary;
+  if (err != CL_SUCCESS) {
+    // The binary might be corrupted (e.g. process killed during write,
+    // or incompatible OpenCL driver update). Just delete the cached
+    // binary silently so we generate a fresh one.
+    unlink(programPath.c_str());
+    return nullptr;
+  }
+  return prog;
+}
+
+void OpenCLFunction::saveProgramToDiskCache(std::string cacheDirectory,
+                                            std::string programFilename,
+                                            cl_program program, cl_context ctx,
+                                            cl_device_id deviceId) {
+
+  std::string programPath = cacheDirectory + '/' + programFilename;
+  if (!llvm::sys::fs::is_directory(cacheDirectory)) {
+    std::error_code err = llvm::sys::fs::create_directories(cacheDirectory);
+    CHECK(!err) << "Could not create the OpenCL program disk cache directory "
+                << cacheDirectory << ".";
+  }
+  cl_uint numDevices = 0;
+  cl_int errC = clGetProgramInfo(program, CL_PROGRAM_NUM_DEVICES,
+                                 sizeof(cl_uint), &numDevices, nullptr);
+  CHECK_EQ(errC, CL_SUCCESS)
+      << "clGetProgramInfo for CL_PROGRAM_NUM_DEVICES failed";
+  CHECK_EQ(numDevices, 1) << "Only one OpenCL device supported";
+
+  size_t binSize = 0;
+  errC = clGetProgramInfo(program, CL_PROGRAM_BINARY_SIZES, sizeof(size_t *),
+                          &binSize, nullptr);
+  CHECK_EQ(errC, CL_SUCCESS)
+      << "clGetProgramInfo for CL_PROGRAM_BINARY_SIZES failed";
+
+  unsigned char *bin = new unsigned char[binSize];
+  errC = clGetProgramInfo(program, CL_PROGRAM_BINARIES, binSize, &bin, nullptr);
+  CHECK_EQ(errC, CL_SUCCESS)
+      << "clGetProgramInfo for CL_PROGRAM_BINARIES failed.";
+
+  int fileDesc = open(programPath.c_str(), O_WRONLY | O_CREAT | O_TRUNC,
+                      S_IRUSR | S_IWUSR);
+
+  CHECK_GT(fileDesc, -1) << "Error opening " << programPath << " for writing.";
+
+  ssize_t numWritten = write(fileDesc, bin, binSize);
+  if (numWritten == -1)
+    perror("Could not write binary:");
```

Comment:
Why perror instead of CHECK_* or some other mechanisms used in Glow?

Reply:
That was an accidental leftover. Removed (in the next update).